xrmocap/model/mvp/projattn.py

```diff
@@ -0,0 +1,281 @@
+            self.logger.warning("You'd better set d_model in Deform to "
+                                'make the dimension of each attention '
+                                'head a power of 2 which is more efficient '
+                                'in our CUDA implementation.')
+            warnings.warn("You'd better set d_model in Deform to "
+                          'make the dimension of each attention '
+                          'head a power of 2 which is more efficient '
+                          'in our CUDA implementation.')
+
+        self.im2col_step = 64
+
+        self.d_model = d_model
+        self.n_levels = n_levels
+        self.n_heads = n_heads
+        self.n_points = n_points
+
+        self.sampling_offsets = nn.Linear(d_model,
+                                          n_heads * n_levels * n_points * 2)
+        self.attention_weights = nn.Linear(d_model,
+                                           n_heads * n_levels * n_points)
+        if projattn_pos_embed_mode == 'use_rayconv':
+            self.rayconv = nn.Linear(d_model + 3, d_model)
+        elif projattn_pos_embed_mode == 'use_2d_coordconv':
+            self.rayconv = nn.Linear(d_model + 2, d_model)
+        elif projattn_pos_embed_mode == 'ablation_not_use_rayconv':
+            self.rayconv = nn.Linear(d_model, d_model)
+        else:
+            raise ValueError('invalid projective attention posembed mode')
+        self.output_proj = nn.Linear(d_model, d_model)
+
+        self._reset_parameters()
+
+        self.projattn_pos_embed_mode = projattn_pos_embed_mode
+
+    def _reset_parameters(self):
+        constant_(self.sampling_offsets.weight.data, 0.)
+        thetas = torch.arange(self.n_heads, dtype=torch.float32) \
+            * (2.0 * math.pi / self.n_heads)
+        grid_init = torch.stack([thetas.cos(), thetas.sin()], -1)
+        grid_init = (grid_init / grid_init.abs().max(-1, keepdim=True)[0])\
+            .view(self.n_heads, 1, 1, 2).\
+            repeat(1, self.n_levels, self.n_points, 1)
+        for i in range(self.n_points):
+            grid_init[:, :, i, :] *= i + 1
+        with torch.no_grad():
+            self.sampling_offsets.bias = nn.Parameter(grid_init.view(-1))
+        constant_(self.attention_weights.weight.data, 0.)
+        constant_(self.attention_weights.bias.data, 0.)
+        xavier_uniform_(self.rayconv.weight.data)
+        constant_(self.rayconv.bias.data, 0.)
+        xavier_uniform_(self.output_proj.weight.data)
+        constant_(self.output_proj.bias.data, 0.)
+
+    def forward(self,
+                query,
+                reference_points,
+                src_views,
+                camera_ray_embeds,
+                input_spatial_shapes,
+                input_level_start_index,
+                input_padding_mask=None):
+        """Forward function for Projective Attention.
+
+        Args:
+            query :
+                (n_views, Length_{query}, C)
+            reference_points :
```

Comment:
Describe the semantic meaning of`reference_points` and `query`?

Reply:
Added in [d5025b7](https://github.com/openxrlab/xrmocap/pull/21/commits/d5025b700c504f02e48cdaae421d19a4d3d2c8d4).